apache/superset
# feat: Add Deck.gl Contour Layer

### SUMMARY


Adds the Contour Layer from deck.gl [here](https://deck.gl/docs/api-reference/aggregation-layers/contour-layer)

Related SIP: https://github.com/apache/superset/discussions/23553

_**Design Decisions:**_

The addition of the Contour layer follows the structure of existing deck.gl layers to maintain readability and consistency within the codebase.

_Structure_:

The directory `superset-frontend/plugins/legacy-preset-chart-deckgl/src/layers/Contour/` contains the Front end structure for the contour layer visualization.

- `controlPanel` configures the form to grab relevant information specified by the user to create and tune the contour layer.
- `Contour.tsx` contains the functions that define the Contour Layer and get relevant data.
- `index.ts` serves as the wrapper for the contour component and provides a metadata function to grab information to describe what this visualization is/does.

The contour layer requires a unique input of "contours". Contours are defined as either `Isolines` or `Isobands` (more information [here](https://deck.gl/docs/api-reference/aggregation-layers/contour-layer)). 
- `Isolines` require three fields, `threshold`, `color`, `strokeWidth`. `threshold` defines a value to separate aggregated values on the Map. `strokeWidth` defines the width of the line in pixels that separates these values. `color` is just the color of that line.
- `Isobands` are similar to `Isolines`, except the `threshold` takes 2 values to create polygons on the map instead of lines.

To account for this unique structure, I added a new control panel input component: 
![Screenshot 2023-05-20 at 11 38 33 PM](https://github.com/apache/superset/assets/36670322/0876a86d-f7f8-4925-a8aa-4889745be575)
The added files for this input component are located under `superset-frontend/src/explore/components/controls/ContourControl/`:

- `types.ts` contains all the types needed for this new component.
- `index.tsx` controls the stateful variables that control adding/editing/removing contours.
- `ContourOption.tsx` displays added contours and an appropriate tooltip.
- `ContourPopoverControl.tsx` defines a popup with appropriate inputs for creating/editing an `isoline` and `isoband`.
- `ContourPopoverTrigger.tsx` wraps the `ContourPopoverControl` and controls when it is displayed.

The Back End code for this feature follows the back end code for other deck.gl visualizations as well, providing functions to structure queried data for the Heatmap component.


### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF


New Visualization Recording:

https://github.com/apache/superset/assets/36670322/6b8da315-23e7-4ed9-a0ce-f0905815a405

<details>

<summary>Changing order of contours</summary>

https://github.com/apache/superset/assets/36670322/118b97f3-7334-4041-9392-5f5171a57996

</details>


### TESTING INSTRUCTIONS


_**Set up**_
1. Start Superset
2. Click on the `Charts` tab at the top of the screen
3. Click the `+ Chart` button at the top right
4. Load a dataset with geo-positional data (longitude/latitude)
5. Search the charts for `Deck.gl Contour` or click on the `Map` chart category
6. Select `deck.gl Contour` and hit `Create New Chart`
7. Update the `Longitude & Latitude` field to the correct columns
8. Select a `Weight` function
9. Select `Create Chart`
10. There should now be a contour layer present on the screen
    - Note: the underlying map component may not be visible if the MAP_BOX_API_KEY is not set. To set it follow the instructions [here](https://superset.apache.org/docs/frequently-asked-questions/#why-is-the-map-not-visible-in-the-geospatial-visualization)
    - Note: It may look like the contour layer is not working if the `Cell Size` parameter is not large enough. For example, using the `flights` dataset, you should increase `Cell Size` to 50000.

Check that the following cause an immediate update
- Check that increasing/decreasing the `Cell Size` control 
- Try adding a new isoline by clicking `Click to add new contour`
- Try adding a new isoband by clicking `Click to add new contour`
- Try clicking an existing contour option and ensure that a popover appears with the appropriate fields pre-filled
- Delet an existing contour option

Other Checks:
- Check that hovering a contour option and check that the tooltip shows appropriate information
- Check that in the contour popover, the submit button is disabled when:
    - Not all fields are filled
    - All fields are filled, but a non-integer is entered in any of the textfields
- Check all descriptions make sense
- Check that changing the order of contours (by drag and drop) updates their z-index.
   - If there are overlapping contours, contours at the top of the list may be hidden by contours beneath them.
   - _Note: z-fighting may occur depending on how many contours are defined/how many are displayed. However this is an issue with the library itself._
 

### ADDITIONAL INFORMATION
